src/include/optimizer/stats/histogram.h

```diff
@@ -0,0 +1,298 @@
+#pragma once
+
+#include <algorithm>
+#include <cassert>
+#include <cfloat>
+#include <cmath>
+#include <cstdint>
+#include <iostream>
+#include <iterator>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "type/value.h"
+#include "type/type.h"
+#include "common/macros.h"
+#include "common/logger.h"
+
+namespace peloton {
+namespace optimizer {
+
+/*
+ * Online histogram implementation based on
+ * http://www.jmlr.org/papers/volume11/ben-haim10a/ben-haim10a.pdf
+ * Specifically Algorithm 1, 3, and 4
+ *
+ * One problem associated with this appraoch is that
+ * int64_t type convert to double might fail, and
+ * calculation may overflow...
+ *
+ * Note: histogram only applys to data that's comparable.
+ *
+ * TOOD:
+ * - move helper functions to private
+ * - Add mean, variance,
+ * - Add thresholds to keep histogram bound constant
+ */
+class Histogram {
+ public:
+
+  class Bin;
+
+  const uint8_t max_bins;   // For performance it shouldn't be greater than 255
+  std::vector<Bin> bins;
```

Comment:
So what is needed is some bins with unique p value, and are always sorted according to the p value. 

Operations are inserting, iterating and merging two bins( modify one and delete another). Will std::set also work? Seems easy to just call insert() for set without manually deciding the place to insert. Not sure which is more efficient.